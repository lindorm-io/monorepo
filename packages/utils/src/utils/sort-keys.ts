import { isObjectLike } from "@lindorm/is";
import { Dict } from "@lindorm/types";

export function sortKeys<T extends Dict = Dict>(arg: T): T {
  const result: Dict = {};

  for (const [key, value] of Object.entries(arg).sort()) {
    if (isObjectLike(value)) {
      result[key] = sortKeys(value);
    } else {
      result[key] = value;
    }
  }

  return result as T;
}
