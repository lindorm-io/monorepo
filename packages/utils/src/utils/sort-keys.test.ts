import { sortKeys } from "./sort-keys";

describe("sortKeys", () => {
  test("should sort object keys alphabetically", () => {
    expect(
      sortKeys({
        e: 1,
        b: 1,
        d: {
          k: 1,
          u: 1,
          a: 1,
          c: 1,
          f: {
            i: 1,
            h: 1,
            g: 1,
          },
        },
        z: 1,
        a: 2,
      }),
    ).toMatchSnapshot();
  });
});
