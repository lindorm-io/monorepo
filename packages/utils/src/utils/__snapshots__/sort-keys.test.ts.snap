// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sortKeys should sort object keys alphabetically 1`] = `
{
  "a": 2,
  "b": 1,
  "d": {
    "a": 1,
    "c": 1,
    "f": {
      "g": 1,
      "h": 1,
      "i": 1,
    },
    "k": 1,
    "u": 1,
  },
  "e": 1,
  "z": 1,
}
`;
